Mathlib/Combinatorics/SimpleGraph/IndependentSet.lean

```diff
@@ -0,0 +1,219 @@
+  · exact (c vins).left rfl
+
+-- The neighbors of a vertex i form an independent set in a trianlge free graph G.
+theorem isIndependentSet_neighborSet_if_triangleFree [DecidableEq α] (h: G.CliqueFree 3) (v : α) :
+    G.IsIndependentSet (G.neighborSet v) := by
+  by_contra nind
+  simp [SimpleGraph.IsIndependentSet, Set.Pairwise] at nind
+  obtain ⟨j, avj, k, avk, _, ajk⟩ := nind
+  exact h {v, j, k} (is3Clique_triple_iff.mpr (by simp [avj, avk, ajk]))
+
+end IndependentSet
+
+/-! ### N-Independent sets -/
+
+
+section NIndependentSet
+
+variable {n : ℕ} {s : Finset α}
+
+/-- An `n`-IndependentSet in a graph is a set of `n` vertices which are pairwise nonadjacent. -/
+structure IsNIndependentSet (n : ℕ) (s : Finset α) : Prop where
+  isIndependentSet : G.IsIndependentSet s
+  card_eq : s.card = n
+
+theorem isNIndependentSet_iff : G.IsNIndependentSet n s ↔ G.IsIndependentSet s ∧ s.card = n :=
+  ⟨fun h ↦ ⟨h.1, h.2⟩, fun h ↦ ⟨h.1, h.2⟩⟩
+
+/-- An independent n-set is an n-clique in the complement graph and vice versa. -/
+theorem isNIndependentSet_iff_isNClique_of_complement :
+    G.IsNIndependentSet n s ↔ Gᶜ.IsNClique n s := by
+  rw [isNIndependentSet_iff, isIndependentSet_iff_isClique_of_complement]
+  simp [isNClique_iff]
+
+instance [DecidableEq α] [DecidableRel G.Adj] {n : ℕ} {s : Finset α} :
+    Decidable (G.IsNIndependentSet n s) :=
+  decidable_of_iff' _ G.isNIndependentSet_iff
+
+end NIndependentSet
+
+/-! ### Graphs without independent sets -/
+
+
+section IndependentSetFree
+
+variable {n : ℕ}
+
+/-- `G.IndependentSetFree n` means that `G` has no `n`-independent sets. -/
+def IndependentSetFree (n : ℕ) : Prop :=
+  ∀ t, ¬G.IsNIndependentSet n t
+
+/-- An graph is 'n'-independent set free iff its complement is n-clique free. -/
+theorem isIndependentSetFree_iff_isCliqueFree_of_complement :
+    G.IndependentSetFree n ↔ Gᶜ.CliqueFree n := by
+  simp [IndependentSetFree, isNIndependentSet_iff_isNClique_of_complement, CliqueFree]
+
+/-- `G.IndependentSetFreeOn s n` means that `G` has no `n`-independent sets contained in `s`. -/
+def IndependentSetFreeOn (G : SimpleGraph α) (s : Set α) (n : ℕ) : Prop :=
+  ∀ ⦃t⦄, ↑t ⊆ s → ¬G.IsNIndependentSet n t
+
+end IndependentSetFree
+
+/-! ### Set of independent sets-/
+
+
+section IndependentSetSet
+
+variable {n : ℕ} {s : Finset α}
+
+/-- The `n`-independent sets in a graph as a set. -/
+def independentSetSet (n : ℕ) : Set (Finset α) :=
+  { s | G.IsNIndependentSet n s }
+
+variable {G}
+
+@[simp]
+theorem mem_independentSetSet_iff : s ∈ G.independentSetSet n ↔ G.IsNIndependentSet n s :=
+  Iff.rfl
+
+end IndependentSetSet
+
+/-! ### Coclique Number-/
+
+
+section CocliqueNumber
+
+variable {α : Type*} {G : SimpleGraph α}
+
+/-- The maximal number of vertices of an independent set in a graph `G`. -/
+noncomputable def cocliqueNum (G : SimpleGraph α) : ℕ := sSup {n | ∃ s, G.IsNIndependentSet n s}
+
+lemma cocliqueNum_eq_compl_cliqueNum : G.cocliqueNum = Gᶜ.cliqueNum := by
```

Comment:
```suggestion
@[simp] lemma cliqueNum_compl : Gᶜ.cliqueNum = G.cocliqueNum := by
```
and can we also have `cocliqueNum_compl`?

Reply:
I tried this just now, but the simpNF linter sees a problem I can't seem to reproduce locally:
```
/- The `simpNF` linter reports:
SOME SIMP LEMMAS ARE NOT IN SIMP-NORMAL FORM.
see note [simp-normal form] for tips how to debug this.
https://leanprover-community.github.io/mathlib_docs/notes.html#simp-normal%20form -/
-- Mathlib.Combinatorics.SimpleGraph.IndependentSet
Error: ././././Mathlib/Combinatorics/SimpleGraph/IndependentSet.lean:139:1: error: @SimpleGraph.cliqueNum_compl LINTER FAILED:
simplify fails on left-hand side:
tactic 'simp' failed, nested error:
maximum recursion depth has been reached
use `set_option maxRecDepth <num>` to increase limit
use `set_option diagnostics true` to get diagnostic information
```
I only found anecdotal descriptions of the normal form in the manuals and am now confused. Ideas?

Reply:
You missed something very important about my suggestion: I am putting the `ᶜ` on the LHS rather than the RHS. Since simp always rewrites left to right, adding both `G.cliqueNum = Gᶜ.cocliqueNum` and `G.cocliqueNum = Gᶜ.cliqueNum` means that it rewrites `G.cocliqueNum` to `Gᶜ.cliqueNum` to `Gᶜᶜ.cocliqueNum` to `Gᶜᶜᶜ.cliqueNum`...

Reply:
Oh my, that is very important indeed! Walking right into a trap with my very first simp lemma :D